src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -201,14 +202,47 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
             return validName;
         }
 
-        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(Type type)
+        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(
+            Type type, 
+            ErrorSink errorSink)
         {
-            var properties = type.GetRuntimeProperties().Where(IsAccessibleProperty);
-            var attributeDescriptors = properties.Select(ToAttributeDescriptor);
+            var accessibleProperties = type.GetRuntimeProperties().Where(IsAccessibleProperty);
+            var attributeDescriptors = new List<TagHelperAttributeDescriptor>();
+
+            foreach (var property in accessibleProperties)
+            {
+                var descriptor = ToAttributeDescriptor(property);
+                if (ValidateTagHelperAttributeDescriptor(descriptor, type, errorSink))
+                {
+                    attributeDescriptors.Add(descriptor);
+                }
+            }
 
             return attributeDescriptors;
         }
 
+        private static bool ValidateTagHelperAttributeDescriptor(
+            TagHelperAttributeDescriptor attributeDescriptor,
+            Type parentType,
+            ErrorSink errorSink)
+        {
+            // data-* attributes are explicitly not implemented by user agents and are not intended for use on 
+            // the server; therefore it's invalid for TagHelpers to bind to them.
+            if (attributeDescriptor.Name.StartsWith(DataDashPrefix, StringComparison.OrdinalIgnoreCase))
+            {
+                errorSink.OnError(
+                    SourceLocation.Zero, 
```

Comment:
Oh I see, I was thinking this was for the consumer of the tag helper, but you're saying it's for the author. Then I think this is fine.